include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,177 @@
+                      }
+                    , indexes_
+                    );
+      Allocator::aligned_dealloc(storage_, size_);
+    }
+
+    EVE_FORCEINLINE auto data_aligned()
+    {
+      auto ptrs = kumi::map([](auto* m) { return as_aligned_pointer(m); }, indexes_);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([](auto const* m) { return as_aligned_pointer(m); }, indexes_);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      std::swap(storage_,other.storage_);
+      std::swap(indexes_,other.indexes_);
+      std::swap(size_,other.size_);
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept
+    {
+      a.swap(b);
+    }
+
+    void build(std::size_t n, Type v = {})
+    {
+      auto sub_span = storage_;
+
+      kumi::for_each( [&]<typename T>(T*& s, auto m)
+                      {
+                        s = reinterpret_cast<T*>(sub_span);
+                        for(std::size_t i=0;i<n;++i) new (s+i) T(m);
+                        sub_span += realign<T>(n);
+                      }
+                    , indexes_, v
+                    );
+
+      capacity_ = available_elements(n);
+    }
+
+    void reindex(std::size_t n)
+    {
+      auto sub_span = storage_;
+
+      kumi::for_each( [&]<typename T>(T*& s)
+                      {
+                        s = reinterpret_cast<T*>(sub_span);
+                        sub_span += realign<T>(n);
+                      }
+                    , indexes_
+                    );
+      capacity_ = available_elements(n);
+    }
+
+    //==============================================================================================
+    // Static helpers
+    //==============================================================================================
+    template<typename T> static EVE_FORCEINLINE auto as_aligned_pointer(T* ptr)
+    {
+      return eve::as_aligned(ptr, eve::detail::cache_line_cardinal<T>{});
+    }
+
+    template<typename T>
+    static  auto realign(std::size_t n) { return eve::align(sizeof(T)*n, eve::over{64}); }
+
+    static auto available_elements(std::size_t n) noexcept
+    {
+      // Capacity is the smallest gap possible, ie the one in the largest type
+      constexpr std::size_t largest = max_scalar_size_v<Type>;
+      return eve::align(largest*n, eve::over{64})/largest;
+    }
+
+    static auto byte_size(std::size_t n) noexcept
+    {
+      return  kumi::fold_left ( [n]<typename S>(auto a, S) { return a + realign<S>(n); }
+                              , flat_t{}
+                              , 0ULL
+                              );
+    }
```

Comment:
I think this function should be called from capacity and not size.

Then it's just sum(size of each field * n);

Reply:
Are we OK the current implementation is correct as per discussion on Slack ?

Reply:
yep